### PR TITLE
Download media through yt-dlp

### LINKED ISSUE #11: Add feature for downloading streamed content
Issue for planned feature.

### PR DESCRIPTION
**Pull request by StellaSmith**
closes #11

### CODE REVIEW

File: src/external.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            .next()
+            .unwrap_or_default();
+        if let Ok(res) = req_client.post(url).send().await {
+            Some((res.bytes(), extension))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
```
Comment (StellaSmith) [inline src/external.rs:205]: Ideally all these downloads should happen in parallel

File: src/external.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    std::fs::DirBuilder::new()
+        .recursive(true)
+        .create(&path)
+        .unwrap();
+    Command::new("yt-dlp")
+        .arg("-P")
+        .arg(path)
+        .args([
+            "--no-playlist",
+            "--write-info-json",
```
Comment (StellaSmith) [inline src/external.rs:253]: You can use ` --dump-single-json` which outputs the whole json info to stdout instead

**Review by StellaSmith — COMMENTED** [review]
Adding `--default-search "ytsearch"` to the args would be nice so the user doesn't has to always specify the exact url

File: Cargo.toml [hunk 3]
```diff
@@ -38,5 +38,8 @@ async-process = "1.5"
 serde = "1.0"
 # For parsing info.json
 serde_json = "1.0"
+
 # For parsing language identifiers
-isolang = "2.0"
+[dependencies.isolang]
+features = ["serde"]
+version = "2.0"
```
Comment (StellaSmith) [inline Cargo.toml:45]:
```suggestion
isolang = { version = "2.0", features = ["serde"] }
```

File: src/main.rs [hunk 4]
```diff
@@ -8,6 +10,15 @@ use serenity::model::channel::Message;
 use serenity::model::id::ChannelId;
 use serenity::prelude::*;
 
+const STREAM_DIR: &str = "./stream_queue/";
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/main.rs]: I made a quick fix for now. Still global, but not thread safe.

File: src/hashing.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    let mut output = String::new();
+    for byte in bytes {
+        output = format!["{output}{:02X}", byte];
+    }
+    output
+}
+
+pub fn hash_url(url: Url) -> UrlHash {
+    UrlHash(encode_hex(sha256_digest(&url.to_string()).as_ref()))
+}
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/hashing.rs]: Removed this and changed the logic.

File: src/external.rs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+use async_process::Command;
+use reqwest::Client;
+use serde_json::{de::from_reader, Value};
+use unix_named_pipe::{create, open_write};
+use url::Url;
+
+use crate::hashing::{hash_url, UrlHash};
+use crate::STREAM_DIR;
+
+const MAX_VIDEO_SIZE: u64 = 1024 * 1024 * 100;
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/external.rs]: Done, but might remove this eventually anyway.

File: src/external.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+const MAX_VERTICAL_RESOLUTION: u64 = 720;
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/external.rs]: Ditto

File: src/external.rs [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    } else if let Some(Value::String(nym)) = data.get("title") {
+        Some(nym.to_owned())
+    } else {
+        None
+    }
+}
+
+pub async fn download_stream(url_hash: UrlHash, req_client: Client) {
+    let base_path = format!["{STREAM_DIR}/{}", url_hash.0];
+    let json_path = format!["{base_path}/info.json"];
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/external.rs]: Done, and sanitized file names.

File: src/external.rs [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    if let Some(Value::String(nym)) = data.get("fulltitle") {
+        Some(nym.to_owned())
+    } else if let Some(Value::String(nym)) = data.get("title") {
+        Some(nym.to_owned())
+    } else {
+        None
+    }
+}
+
+pub async fn download_stream(url_hash: UrlHash, req_client: Client) {
```
[the remark this answers is not in the document]
Comment (elfein727) [reply src/external.rs]: Changed this to use platform and video name.

File: src/external.rs (hunk 1 above)
Comment (elfein727) [reply src/external.rs]: Should be fixed in new changes.

File: src/external.rs (hunk 2 above)
Comment (elfein727) [reply src/external.rs]: Ended up doing this. The shorthand is `-j`.

**Comment by elfein727** [comment]
> Adding `--default-search "ytsearch"` to the args would be nice so the user doesn't has to always specify the exact url

I'll do that soon- basically just do search on URL parse fail?

File: Cargo.toml (hunk 3 above)
Comment (elfein727) [reply Cargo.toml]: Pushed.

File: src/downloading/yt_dlp.rs [hunk 10]
```diff
@@ -17,6 +17,8 @@ pub async fn download_info_json(url: Url) -> StreamData {
             "\"(bv[height<=720]+ba/bv[height<=720]*+ba/b[height<=720])[filesize<100M]\"",
             // Reject playlists
             "--no-playlist",
+            // Make YT default
+            "--default-search \"ytsearch\"",
```
Comment (StellaSmith) [inline src/downloading/yt_dlp.rs:21]:
```diff
-"--default-search \"ytsearch\"",
+"--default-search", "ytsearch",
```

File: Cargo.toml [hunk 11]
```diff
@@ -19,6 +19,7 @@ env-file-reader = "0.3"
 
 lazy_static = "1.4"
 log = "0.4"
+<<<<<<< HEAD
```
Comment (StellaSmith) [inline Cargo.toml]: this is not valid

File: src/downloading/yt_dlp.rs [hunk 12]
```diff
@@ -18,7 +18,8 @@ pub async fn download_info_json(url: Url) -> StreamData {
             // Reject playlists
             "--no-playlist",
             // Make YT default
-            "--default-search \"ytsearch\"",
+            "--default-search",
+            "\"ytsearch\"",
```
Comment (StellaSmith) [inline src/downloading/yt_dlp.rs:22]:
```diff
-"\"ytsearch\"",
+"ytsearch",

**Pull request merged**